observer.py

```diff
@@ -0,0 +1,311 @@
+from __future__ import annotations
+import operator
+import os
+import json
+import time
+import logging
+import logging.config
+from typing import Optional
+import asyncio
+
+from discord_webhook import DiscordEmbed, DiscordWebhook
+
+from data_classes import EmbedData, TokenSale
+from api_calls import ApiCalls
+
+
+with open("configs/log_config.json", "r", encoding="UTF-8") as stream:
+    config = json.load(stream)
+logging.config.dictConfig(config)
+logger = logging.getLogger("standard")
+with open("configs/config.json", encoding="UTF-8") as g:
+    configs = json.load(g)
+
+
+class SaleFinderSubject:
+    """
+    Manages observers and filter. Gets sales data, processes it and notifies observers.
+    """
+
+    token_sale_list: list[TokenSale]
+    raw_sales_data: dict = {}
+    avax_price_in_usd: float = 0
+    sale_to_notify: TokenSale
+    embed_data: EmbedData
+    _observers: list[FilteredDiscordObserver] = []
+    _observed_collections: list[str] = []
+    _last_notified_transactions: list[str]
+
+    def __init__(self) -> None:
+        with open("configs/lastNotifiedTransactions.json", encoding="UTF-8") as file:
+            self._last_notified_transactions = json.load(file)
+        self.JOEPEGS_API_KEY: Optional[str] = os.getenv("JOEPEGS_API_KEY")
+
+    def attach(self, observer: FilteredDiscordObserver) -> None:
+        self._observers.append(observer)
+        logger.debug("Subject: Attached an observer.")
+
+    def detach(self, observer: FilteredDiscordObserver) -> None:
+        self._observers.remove(observer)
+        logger.debug("Subject: Detached an observer.")
+
+    def set_filter(self, collection_ids: list[str]) -> None:
+        self._observed_collections = [
+            collection.lower() for collection in collection_ids
+        ]
+        logger.warning(
+            "Filter was set up. Will only notify about these collections: %s",
+            self._observed_collections,
+        )
+
+    async def run(self) -> None:
+        """
+        Get sales data, process it, notify observers and save notified.
+        """
+        logger.debug("Run method executing")
+        self.api_calls = ApiCalls()
+        self.raw_sales_data: list[
+            dict
+        ] = await self.api_calls.ask_joepegs_about_recent_sales()
+
+        if self.new_sales():
+            logger.debug("New sales were found")
+            avax_price_in_usd = await self.api_calls.ask_thegraph_avax_price()
+            if avax_price_in_usd:
+                self.avax_price_in_usd = avax_price_in_usd
+            self.remove_notified_sales()
+            await self.get_sales_data_from_joepegs()
+            self.filter_and_notify()
+            self.write_notified_sales_to_file()
+        else:
+            logger.debug("New sales were not found")
+
+    def new_sales(self):

```

Comment:
small nit but in general, private instance methods that aren't intended to be used by external callers should be prefixed with underscore (e.g. `def _new_sales(self):`)